### ISSUE TITLE
Did you try to reproduce the results with the same hyperparameters from the NASBench101 paper?

### ISSUE THREAD

**Issue by longerHost** [romulus0914/NASBench-PyTorch#6]
Hi @gabikadlecova , I am wondering did you try to reproduce the training results from the NASBench101 paper with the exact same hyperparameters?

I found their hyperparameter setting in the Supplementary Material section at the end of the paper.
![7fab4b45e0a7a31e20c74d3ab7b9eef](https://user-images.githubusercontent.com/8159957/199595821-0dfaba8a-8298-40b6-8781-4b7dd5d89cde.png)

The strange thing is that when I applied their hyperparameters to our code here(batchsize=256, optimizer=RMSProp, etc.) the training accuracy dropped to around 0.6-0.7,  I query the results by the paper's repo `api`, the acc should be around 0.93-0.94.

Also, I am wondering why the default hyperparameter setting is different from the official repo? Like [batchsize](https://github.com/romulus0914/NASBench-PyTorch/blob/master/main.py#L52) and [optimizer](https://github.com/romulus0914/NASBench-PyTorch/blob/master/main.py#L87).

And, thanks for your response as always.

**Comment by gabikadlecova** [comment]
Hi,
1. did you try to reproduce the training results from the NASBench101 paper with the exact same hyperparameters?
I unfortunately wasn't able to do that, because I don't have the necessary resources. When training with the default hyperparameters, I was able to get different results for different architectures, some with accuracy over 0.9, some below, so I used that for my work.

Did you train it for the original 108 epochs? And the accuracy stayed so low?
I realized that the batch normalization hyperparameters can't be set easily (PyTorch has different default values). I will push an update that enables to set them easily.

2. Why the default hyperparameter setting is different from the official repo
The values were chosen by the author of the repo (@romulus0914 ), I did not change them since they worked well for me. If we manage to reproduce the original hyperparameters, I can change the defaults.

**Comment by gabikadlecova** [comment]
The update is pushed - batch norm hyperparameters can be set from main.py (as arguments of the Network class).

I'm quite busy right now, but I'll ask if my institution has a gpu with a large enough memory size to test the original hyperparameters.

**Comment by longerHost** [comment]
Thanks for your reply, I will help run the new models and let you know the results. 

> Did you train it for the original 108 epochs? And the accuracy stayed so low?

Yes, for the previous runs, I trained 120 epochs with the parameters from `NASBench101` paper and the `acc` is low.

> I realized that the batch normalization hyperparameters can't be set easily (PyTorch has different default values). 

This is very helpful and I might not properly set these two parameters last time, I will try it again.

>  If we manage to reproduce the original hyperparameters, I can change the defaults.

Thanks, let's see whether we could make it and it will be great if we are able to reproduce the result from the paper.

**Comment by gabikadlecova** [comment]
Thanks a lot for trying to run it, hopefully we can reproduce the results. In a week or two I should have more time, so I'll try to compare the original implementation with this one, and look into differences in Tensorflow vs PyTorch - e.g. I'm not sure if they have the same default layer initializations.

**Comment by gabikadlecova** [comment]
@longerHost so I started looking into the original nasbench code and it seems like the key is to set RMSProp eps to 1.0. I added an argument to main: `--rmsprop_eps 1.0`. When I ran it with the smaller batch size, the accuracy went up to 0.84 after only 5 epochs compared to <0.8 accuracy after the 100 epochs. You can find the modified code in branch `tf_like`,

Also, it turns out PyTorch uses a different implementation of RMSProp than Tensorflow - on equation level! See [here](https://github.com/pytorch/pytorch/issues/32545) and [here](https://github.com/pytorch/pytorch/issues/23796).
As suggested in the comments, I added [RMSpropTF](https://github.com/rwightman/pytorch-image-models/blob/main/timm/optim/rmsprop_tf.py#L5) as an optimizer - an implementation of RMSProp that matches the Tensorflow variant. It's part of the [pytorch-image-models](https://github.com/rwightman/pytorch-image-models) repo, installable by `pip install timm`. So you can try that as well: `--optimizer rmsprop_tf`.

Finally, I modified the code so that Conv2D and Linear layers have the same weight initializations as in Tensorflow, you can turn that on with `--tf_like True`.

The exact hyperparameters I used:
```
python main.py --batch_norm_momentum 0.997 --batch_norm_eps 1.0 --optimizer rmsprop --learning_rate 0.2  --rmsprop_eps 1.0
```

**Comment by longerHost** [comment]
Thanks for the update, it is so interesting to know that the implementation of `RMSProp` between TensorFlow and PyTorch is different.

I use the same command in your reply and finish the 100 epochs training, the result looks pretty good,
`Final test metrics: {'test_loss': 0.2139013260602951, 'test_accuracy': 0.9442999958992004}`

To compare with the previous work, I directly run `python main.py` and the results is:
`Final test metrics: {'test_loss': 0.4304623305797577, 'test_accuracy': 0.9101999998092651}`

Then, I tested `batch_size 256` at epoch `108` with the command `python main.py --batch_norm_momentum 0.997 --batch_norm_eps 1.0 --optimizer rmsprop --learning_rate 0.2  --rmsprop_eps 1.0 --batch_size 256 --epochs 108`

The result is:
`Final test metrics: {'test_loss': 0.2612570524215698, 'test_accuracy': 0.939799964427948}`

To train the TensorFlow style `rmsprop` with the command `python main.py --batch_norm_momentum 0.997 --batch_norm_eps 1.0 --optimizer rmsprop_tf --learning_rate 0.2  --rmsprop_eps 1.0 --tf_like True --batch_size 256 --epochs 108` which I believe this is same with the paper's original setting,

The result is:
`Final test metrics: {'test_loss': 0.2576700448989868, 'test_accuracy': 0.9380999803543091}`

All the training is based on the code from `tf_like` branch.

I will have more time to train some other `specs` around this Friday and will compare them with the `api` results from the paper's repo.

**Comment by gabikadlecova** [comment]
Hi @longerHost,
thanks for the accuracies! It's much better than before, seems like the problem is solved.

I have gone through the whole code and compared it with the TensorFlow version. Both code and hyperparameters seem to match, but I realized that the validation set in our implementation is randomly sampled, but in their code they use a fixed part of the training set as the validation set. I created an [issue](https://github.com/romulus0914/NASBench-PyTorch/issues/7) for this, and will fix that tommorow.

Maybe wait with the `api` test until I fix the validation set so that we have more accurate results, but if you've already started, it will be probably quite similar. The results won't match 100 % anyway, since the augmentation won't be the same, and due to inherent differences in TensorFlow and PyTorch implementations.

**Comment by gabikadlecova** [comment]
Hi @longerHost,

I fixed the validation set, so now the models are trained on the same train-val-test splits as in the original paper. I pushed the fix to `tf_like`.

So after we verify that we can reproduce the original hyperparams, I will set them as defaults (including e.g. the batch size) and merge `tf_like` into `master`.

**Comment by longerHost** [comment]
Hi @gabikadlecova thanks for your reply, I got Covid last week during a Thanksgiving short travel, and I didn't get time to work on this. I think the results look much better than before and I agree with you that we don't need to get the exact same results with them due to lots of randomnesses, only if the results are reasonable will be good enough. I will try training and verifying the models with the new code this weekend. 

Could be earlier than that, I will post the results here once I got it done.

**Comment by gabikadlecova** [comment]
Hi @longerHost, no worries - get well soon, I hope Covid isn't too serious for you.

**Comment by longerHost** [comment]
Hi @gabikadlecova , I already recovered and it was not very serious for me,  thanks.

I have finished the training for 8 top acc models and 8 randomly selected models, the results at the 108th epoch of the `trained` and `API` are listed in the following tables.

The training command I used:
`python main.py --batch_norm_momentum 0.997 --batch_norm_eps 1.0 --optimizer rmsprop --learning_rate 0.2  --rmsprop_eps 1.0 --batch_size 256 --epochs 120`

Table 1, top acc models comparison:
| Model | Trained | API |
| -----| ---- | ---- |
| model0 | 0.9432999491691589 | 0.9449118375778198 |
| model1 | 0.9521999955177307 | 0.9447115659713745 |
| model2 | 0.9380999803543091| 0.9446113705635071 |
| model3 | 0.9432999491691589| 0.9455128312110901 |
| model4 | 0.9454999566078186 | 0.9515224099159241 |
| model5 | 0.9422000050544739 | 0.9487179517745972 |
| model6 | 0.94159996509552| 0.9443109035491943 |
| model7 | 0.9438999891281128 | 0.9463140964508057 |

Table 2, random models comparison:
| Model | Trained | API |
| -----| ---- | ---- |
| model0 | 0.8906999826431274 | 0.8840144276618958 |
| model1 | 0.930899977684021 | 0.932692289352417 |
| model2 | 0.914900004863739 | 0.9106570482254028 |
| model3 | 0.9307999610900879 | 0.924879789352417 |
| model4 | 0.8848999738693237 | 0.8793069124221802 |
| model5 | 0.9136999845504761 | 0.905348539352417 |
| model6 | 0.9185000061988831 | 0.9102563858032227 |
| model7 | 0.9143999814987183 | 0.8879206776618958 |

**Due to the randomness, it is almost impossible to fully reproduce the results and make them the same as the original paper, but based on the results we verified, I think it is obvious that we could argue the nasbench101 results in this pytroch-based repo were reproduced.**

More details, for the specific models I used:

top models:
```
def Load_tops_matix_and_ops():
    matrix_list = []
    operations_list = []

    # pickle specs no filter file pickle_spec0
    #test_accuracy is 94.49118375778198 total num large than 94 is 6 now
    matrix0 = [[0, 1, 0, 0, 1, 1],
    [0, 0, 1, 1, 1, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0]]
    operations0 = ['input', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'output']

    matrix_list.append(matrix0)
    operations_list.append(operations0)

    # pickle specs no filter file pickle_spec1
    # test_accuracy is 94.47115659713745 total num large than 94 is 4 now
    matrix1 = [[0, 1, 1, 1, 1],
    [0, 0, 1, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0]]  
    operations1 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'output']

    matrix_list.append(matrix1)
    operations_list.append(operations1)

    # pickle specs no filter file pickle_spec2
    # test_accuracy is 94.46113705635071 total num large than 94 is 5 now
    matrix2 = [[0, 1, 1, 0, 0, 1, 1],
    [0, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 0, 0],
    [0, 0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0]]
    operations2 = ['input', 'maxpool3x3', 'conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'output']
    matrix_list.append(matrix2)
    operations_list.append(operations2)

    # pickle specs no filter file pickle_spec3
    #test_accuracy is 94.55128312110901 total num large than 94 is 3 now
    matrix3 = [[0, 1, 1, 0, 1, 1],
    [0, 0, 1, 1, 0, 0],
    [0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0]]
    operations3 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'output']

    matrix_list.append(matrix3)
    operations_list.append(operations3)

    # 0.9515224099159241
    matrix4 = [[0, 1, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0]]
    operations4 =  ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'output']
    matrix_list.append(matrix4)
    operations_list.append(operations4)

    # 0.9487179517745972
    matrix5 = [[0, 1, 1, 0, 0, 1, 1],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0]]
    operations5 = ['input', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'output']
    matrix_list.append(matrix5)
    operations_list.append(operations5)

    # 94.43109035491943
    matrix6 = [[0, 1, 1, 1, 1],
       [0, 0, 1, 1, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0]]
    operations6 = ['input', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'output']
    matrix_list.append(matrix6)
    operations_list.append(operations6)

    # 0.9463140964508057
    matrix7 = [[0, 1, 1, 1, 0, 0, 1],
       [0, 0, 1, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0]]
    operations7 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'conv1x1-bn-relu', 'output']
    matrix_list.append(matrix7)
    operations_list.append(operations7)

    # 0.9479166865348816
    matrix8 = [[0, 1, 1, 1, 0, 0, 1],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1, 0, 0],
       [0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 0, 0, 0]]
    operations8 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'output']
    
    matrix_list.append(matrix8)
    operations_list.append(operations8)
    return matrix_list, operations_list
```
random models:
```
def Load_random_matix_and_ops():
    matrix_list = []
    operations_list = []

    # spec 1
    matrix0 = [[0, 1, 0, 0, 1, 0, 0],
        [0, 0, 1, 0, 0, 1, 0],
        [0, 0, 0, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations0 = ['input', 'maxpool3x3', 'conv1x1-bn-relu', 'maxpool3x3', 'conv1x1-bn-relu', 'conv1x1-bn-relu', 'output']
    matrix_list.append(matrix0)
    operations_list.append(operations0)

    # spec 2
    matrix1 = [[0, 1, 1, 1, 0, 1, 0],
            [0, 0, 0, 0, 0, 0, 1],
            [0, 0, 0, 0, 0, 0, 1],
            [0, 0, 0, 0, 1, 0, 0],
            [0, 0, 0, 0, 0, 0, 1],
            [0, 0, 0, 0, 0, 0, 1],
            [0, 0, 0, 0, 0, 0, 0]]
    operations1 = ['input', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'maxpool3x3', 'output']
    matrix_list.append(matrix1)
    operations_list.append(operations1)

    # spec3
    matrix2 = [[0, 1, 1, 1, 0, 0, 0],
    [0, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0]]
    operations2 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'maxpool3x3', 'maxpool3x3', 'maxpool3x3', 'output']
    matrix_list.append(matrix2)
    operations_list.append(operations2)

    # spec4
    matrix3 = [[0, 1, 0, 0, 1, 1, 0],
        [0, 0, 1, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations3 = ['input', 'conv3x3-bn-relu', 'conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'output']
    matrix_list.append(matrix3)
    operations_list.append(operations3)

    # spec5
    matrix4 = [[0, 1, 0, 1, 0, 0, 0],
        [0, 0, 1, 1, 0, 0, 0],
        [0, 0, 0, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations4 = ['input', 'conv1x1-bn-relu', 'maxpool3x3', 'conv1x1-bn-relu', 'maxpool3x3', 'conv1x1-bn-relu', 'output']
    matrix_list.append(matrix4)
    operations_list.append(operations4)

    # spec6
    matrix5 = [[0, 1, 0, 0, 1, 0, 0],
        [0, 0, 1, 0, 1, 0, 0],
        [0, 0, 0, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations5 = ['input', 'maxpool3x3', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'output']
    matrix_list.append(matrix5)
    operations_list.append(operations5)

    # spec7
    matrix6 = [[0, 1, 1, 0, 0, 0, 0],
        [0, 0, 1, 1, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 1, 0, 0],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations6 = ['input', 'conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'output']
    matrix_list.append(matrix6)
    operations_list.append(operations6)

    # spec8
    matrix7 = [[0, 1, 0, 0, 0, 0, 0],
        [0, 0, 1, 1, 1, 0, 1],
        [0, 0, 0, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0]]
    operations7 = ['input', 'conv3x3-bn-relu', 'conv1x1-bn-relu', 'conv3x3-bn-relu', 'maxpool3x3', 'maxpool3x3', 'output']
    matrix_list.append(matrix7)
    operations_list.append(operations7)
    return matrix_list, operations_list

```

The detailed training log is [here](https://drive.google.com/drive/folders/1QguPSGaTX_Dqt5zsc-quLQM1BZeQ8qE7?usp=sharing).

**Comment by gabikadlecova** [comment]
Hi @longerHost,
thank you for your amazing work! I agree that the results look good, there are still some differences, but in the api the accuracies
differ across seeds, so it should be okay. I also noticed that you set `--batch_norm_eps 1.0` - this `eps` is actually `1e-5`, but that probably didn't influence the results too much.

I set the default hyperparameters to paper defaults, added a section about reproducibility to the readme & added you to Contributors.md. I'll merge `tf_like` to `master` and create a new version.

Thank you again for you help!
